rkt/cp.go

```diff
@@ -0,0 +1,281 @@
+		return "", path
+	}
+
+	parts := strings.SplitN(path, ":", 2)
+	if len(parts) == 1 {
+		return "", path
+	}
+
+	return parts[0], parts[1]
+}
+
+// calculateFullPath returns a validated path.
+// If uuid is empty the provided path is validated and returned as is.
+// If uuid is not empty, the stage2rootfs is calculate and the path is joined to stage2rootfs.
+func calculateFullPath(uuid string, path string, expectDir bool) (string, error) {
+	if uuid == "" {
+		return validatePath(path, expectDir)
+	}
+
+	p, err := pkgPod.PodFromUUIDString(getDataDir(), uuid)
+	if err != nil {
+		return "", errwrap.Wrap(errors.New("problem retrieving pod"), err)
+	}
+	defer p.Close()
+
+	if p.State() != pkgPod.Running {
+		return "", fmt.Errorf("pod %q isn't currently running", p.UUID)
+	}
+
+	appName, err := getAppNameForCp(p)
+	if err != nil {
+		return "", errwrap.Wrap(errors.New("unable to determine app name"), err)
+	}
+
+	stage1RootFs := filepath.Join(p.Path(), "stage1/rootfs")
+	stage2RootFs := filepath.Join(stage1RootFs, common.RelAppRootfsPath(*appName))
+
+	return validatePath(filepath.Join(stage2RootFs, path), expectDir)
+}
+
+// validatePath checks whether the path exists and cleans the path.
+// If checkDir is true, it also checks whether the path points to a directory.
+func validatePath(path string, checkDir bool) (string, error) {
+	cleanPath := filepath.Clean(path)
+
+	info, err := os.Stat(cleanPath)
+	if err != nil {
+		return cleanPath, fmt.Errorf("path not found")
+	}
+
+	if checkDir && !info.IsDir() {
+		return cleanPath, fmt.Errorf("path is not a directory")
+	}
+
+	return cleanPath, nil
+}
+
+// getAppNameForCp returns the app name to enter
+// If one was supplied in the flags then it's simply returned
+// If the PM contains a single app, that app's name is returned
+// If the PM has multiple apps, the names are printed and an error is returned
+func getAppNameForCp(p *pkgPod.Pod) (*types.ACName, error) {
+	if flagCpAppName != "" {
+		return types.NewACName(flagCpAppName)
+	}
+
+	// figure out the app name, or show a list if multiple are present
+	_, m, err := p.PodManifest()
+	if err != nil {
+		return nil, errwrap.Wrap(errors.New("error reading pod manifest"), err)
+	}
+
+	switch len(m.Apps) {
+	case 0:
+		return nil, fmt.Errorf("pod contains zero apps")
+	case 1:
+		return &m.Apps[0].Name, nil
+	default:
+	}
+
+	stderr.Print("pod contains multiple apps:")
+	for _, ra := range m.Apps {
+		stderr.Printf("\t%v", ra.Name)
+	}
+
+	return nil, fmt.Errorf("specify app using \"rkt cp --app= ...\"")
+}
+
+// copyFileOrDirectory copies a source file or directory to a destination directory.
+func copyFileOrDirectory(srcPath string, destDirPath string) error {
+	_, srcFilename := filepath.Split(srcPath)
+	destPath := filepath.Join(destDirPath, srcFilename)
+
+	return copyTree(srcPath, destPath)
+}
+
+// copyTree recursively copies a directory tree, preserving permissions.
+// Symlinks are ignored and skipped.
```

Comment:
This behaviour seems surprising.

If I do `rkt cp $uuid:/path/to/symlink /tmp/` then I wont' get an error, it'll exit success, but nothing will have been copied